src/document/position.js

```diff
@@ -171,7 +203,196 @@ CKEDITOR.define( [ 'utils', 'ckeditorerror' ], ( utils, CKEditorError ) => {
 
 			return Position.createFromParentAndOffset( node.parent, node.getIndex() + 1 );
 		}
+
+		/**
+		 * Checks whether this position is before or after given position.
+		 *
+		 * @param {document.Position} otherPosition Position to compare with.
+		 * @returns {Number} A flag indicating whether this position is {@link #BEFORE} or
+		 * {@link #AFTER} or {@link #SAME} as given position. If positions are in different roots,
+		 * {@link #DIFFERENT} flag is returned.
+		 */
+		compareWith( otherPosition ) {
+			if ( this.root != otherPosition.root ) {
+				return DIFFERENT;
+			}
+
+			const result = utils.compareArrays( this.path, otherPosition.path );
+
+			switch ( result ) {
+				case utils.compareArrays.SAME:
+					return SAME;
+
+				case utils.compareArrays.PREFIX:
+					return BEFORE;
+
+				case utils.compareArrays.EXTENSION:
+					return AFTER;
+
+				default:
+					if ( this.path[ result ] < otherPosition.path[ result ] ) {
+						return BEFORE;
+					} else {
+						return AFTER;
+					}
+			}
+		}
+
+		/**
+		 * Returns a new position that is a combination of this position and given positions. The combined
+		 * position is this position transformed by moving a range starting at `from` to `to` position.
+		 * It is expected that `original` position is inside the moved range.
+		 *
+		 * In other words, this method in a smart way "cuts out" `from` path from this position and
+		 * injects `to` path in it's place, while doing necessary fixes in order to get a correct path.
+		 *
+		 * Example:
+		 * 	let original = new Position( [ 2, 3, 1 ], root );
+		 * 	let from = new Position( [ 2, 2 ], root );
+		 * 	let to = new Position( [ 1, 1, 3 ], otherRoot );
+		 * 	let combined = original.getCombined( from, to );
+		 * 	// combined.path is [ 1, 1, 4, 1 ], combined.root is otherRoot
+		 *
+		 * Explanation:
+		 * We have a position `[ 2, 3, 1 ]` and move some nodes from `[ 2, 2 ]` to `[ 1, 1, 3 ]`. The original position
+		 * was inside moved nodes and now should point to the new place. The moved nodes will be after
+		 * positions `[ 1, 1, 3 ]`, `[ 1, 1, 4 ]`, `[ 1, 1, 5 ]`. Since our position was in the second moved node,
+		 * the transformed position will be in a sub-tree of a node at `[ 1, 1, 4 ]`. Looking at original path, we
+		 * took care of `[ 2, 3 ]` part of it. Now we have to add the rest of the original path to the transformed path.
+		 * Finally, the transformed position will point to `[ 1, 1, 4, 1 ]`.
+		 *
+		 * @param {document.Position} from Beginning of the moved range.
+		 * @param {document.Position} to Position where the range is moved.
+		 * @returns {document.Position} Combined position.
+		 */
+		getCombined( from, to ) {
```

Comment:
`getCombined` is messy: `from` is a begging of the range, but there is not end of that range. `to` is a position (not the end of that range as one could assume). We assume that the position is inside the range, so the function can be called only in specific cases. It looks more like the helper, not the end-user API. It could be moved to the `transformoperation` and changed to be an internal helper or it could be fixed: the first parameter should be a range and the function should check if he position is inside that range. In this case it could be renamed to `getTransformedByMove` and stay here.